lib/storage/src/content_manager/snapshots/recover.rs

```diff
@@ -0,0 +1,181 @@
+use std::path::Path;
+
+use collection::collection::Collection;
+use collection::config::CollectionConfig;
+use collection::operations::snapshot_ops::SnapshotRecover;
+use collection::shards::replica_set::ReplicaState;
+use collection::shards::shard_config::ShardType;
+use collection::shards::shard_versioning::latest_shard_paths;
+
+use crate::content_manager::snapshots::download::{download_snapshot, downloaded_snapshots_dir};
+use crate::{StorageError, TableOfContent};
+
+pub async fn do_recover_from_snapshot(
+    toc: &TableOfContent,
+    collection_name: &str,
+    source: SnapshotRecover,
+) -> Result<bool, StorageError> {
+    let SnapshotRecover { location } = source;
+
+    let snapshot_download_path = downloaded_snapshots_dir(toc.snapshots_path());
+    tokio::fs::create_dir_all(&snapshot_download_path).await?;
+
+    log::debug!(
+        "Downloading snapshot from {} to {}",
+        location,
+        snapshot_download_path.display()
+    );
+
+    let snapshot_path = download_snapshot(location, &snapshot_download_path).await?;
+
+    log::debug!("Snapshot downloaded to {}", snapshot_path.display());
+
+    let tmp_collection_dir = Path::new(toc.storage_path())
+        .join("tmp_collections")
+        .join(collection_name);
+
+    log::debug!(
+        "Recovering collection {} from snapshot {}",
+        collection_name,
+        snapshot_path.display()
+    );
+
+    if tmp_collection_dir.exists() {
+        tokio::fs::remove_dir_all(&tmp_collection_dir).await?;
+    }
+    tokio::fs::create_dir_all(&tmp_collection_dir).await?;
+
+    log::debug!("Unpacking snapshot to {}", tmp_collection_dir.display());
+
+    // Unpack snapshot collection to the target folder
+    Collection::restore_snapshot(&snapshot_path, &tmp_collection_dir)?;
+
+    let snapshot_config = CollectionConfig::load(&tmp_collection_dir)?;
+
+    let collection = toc.get_collection(collection_name).await?;
```

Comment:
maybe this `get_collection` should be the first line here so that we do not touch the file system if the collection does not exist